src/ai/ai.cpp

```diff
@@ -12,6 +12,231 @@
+	auto status = target_is_gp ? fatten(state.world, target_is_gp).get_status() : fatten(state.world, from_is_gp).get_status();
+
+	int32_t level = 0;
+	switch(nations::influence::level_mask & status) {
+		case nations::influence::level_in_sphere:
+			level = 3;
+			break;
+		case nations::influence::level_friendly:
+			level = 2;
+			break;
+		case nations::influence::level_cordial:
+			level = 1;
+			break;
+		case nations::influence::level_neutral:
+			level = 0;
+			break;
+		case nations::influence::level_opposed:
+			level = -1;
+			break;
+		case nations::influence::level_hostile:
+			level = -2;
+			break;
+	}
+	return float(level) * 10.0f;
+}
+
+float sphere_of_influence(sys::state& state, dcon::nation_id target, dcon::nation_id from) {
+	auto t_sphere = fatten(state.world, target).get_in_sphere_of();
+	auto f_sphere = fatten(state.world, from).get_in_sphere_of();
+
+	if(nations::in_same_sphere(state, target, from)) {
+		return 25.0f; // same sphere
+	} else if(nations::is_great_power(state, target) && f_sphere && f_sphere != target) {
+		return -25.0f; // target is GP and from is in a different sphere
+	} else if(nations::is_great_power(state, from) && t_sphere && t_sphere != from) {
+		return -1000.0f; // from is GP and target is in a different sphere
+	} else if(t_sphere && f_sphere) {
+		return -25.0f; // both non-GP and in different spheres
+	}
+	return 0.0f;
+}
+
+float infamy(sys::state& state, dcon::nation_id target, dcon::nation_id from) {
+	return -std::floor(state.world.nation_get_infamy(from) * 2.0f);
+}
+
+float ai_political_considerations(sys::state& state, dcon::nation_id target, dcon::nation_id from) {
+	float score = 0.0f;
+	auto t_rival = state.world.nation_get_ai_rival(target);
+	auto f_rival = state.world.nation_get_ai_rival(from);
+	// bonus if same rivals
+	if(t_rival && t_rival == f_rival)
+		score += 25.0f;
+	if(ai_has_mutual_enemy(state, from, target))
+		score += 25.0f;
+	if(state.world.nation_get_ai_is_threatened(target))
+		score += 25.0f;
+	// TODO: does `from` border `target`'s rival? beneficial for `from`
+	return score;
+}
+
+float civilization_difference(sys::state& state, dcon::nation_id target, dcon::nation_id from) {
+	auto t_status = state.world.nation_get_is_civilized(target);
+	auto f_status = state.world.nation_get_is_civilized(from);
+	if(t_status && !f_status) { // requester unciv, recipient civ
+		return -80.0f;
+	} else if(!t_status && f_status) { // requester civ, recipient unciv
+		return -30.0f;
+	} else {
+		return 0.0f;
+	}
+}
+
+float current_wars(sys::state& state, dcon::nation_id target, dcon::nation_id from) {
+	float value = 0.0f;
+	std::vector<dcon::nation_id> allies;
+	nations::get_allies(state, target, allies);
+	for(auto& a : allies) {
+		if(military::find_war_between(state, from, a)) {
+			value -= 25.0f; // -25/`target` ally that `from` is at war with
+		}
+	}
+
+	if(military::are_in_common_war(state, from, target) && !military::are_at_war(state, from, target)) {
```

Comment:
you may want `military::are_allied_in_war` here -- that is true if they are on the same side of some ongoing war